ruby/src/IceRuby/DefaultSliceLoader.cpp

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) ZeroC, Inc.
+
+#include "DefaultSliceLoader.h"
+#include "Types.h"
+#include "Util.h"
+
+using namespace std;
+using namespace IceRuby;
+
+Ice::SliceLoaderPtr
+IceRuby::DefaultSliceLoader::instance()
+{
+    static Ice::SliceLoaderPtr instance{new DefaultSliceLoader()};
+    return instance;
+}
+
+Ice::ValuePtr
+IceRuby::DefaultSliceLoader::newClassInstance(string_view typeId) const
+{
+    // When the type ID is ::Ice::Object, it indicates that the stream has not found a factory and is providing us an
+    // opportunity to preserve the object.
+    ClassInfoPtr info =
+        typeId == Ice::Value::ice_staticId() ? lookupClassInfo("::Ice::UnknownSlicedValue") : lookupClassInfo(typeId);
+
+    if (!info)
+    {
+        return nullptr;
+    }
+
+    volatile VALUE obj = callRuby(rb_class_new_instance, 0, reinterpret_cast<VALUE*>(0), info->rubyClass);
```

Comment:
volatile is not necessary here. The VALUE is passed to ValueReader which register its address to prevent GC.

The volatile keyword is required in some situations to prevent GC, when variables can be optimized away. See "`volatile` to protect from GC" in https://ruby-hacking-guide.github.io/gc.html

```suggestion
    VALUE obj = callRuby(rb_class_new_instance, 0, reinterpret_cast<VALUE*>(nullptr), info->rubyClass);
```

Reply:
The compiler does not like the nullptr in `reinterpret_cast<VALUE*>(nullptr)`, so I changed it back to 0.